databroker/core.py

```diff
@@ -175,23 +180,29 @@ def get_events(mds, fs, headers, fields=None, stream_name=ALL, fill=False,
             extra_fields = set()
             if fields:
                 event_fields = set(descriptor['data_keys'])
-                discard_fields = event_fields - fields
-                extra_fields = fields - event_fields
+                selected_fields = set(filter(comp_re.fullmatch, event_fields))
+                discard_fields = event_fields - selected_fields
             for event in mds.get_events_generator(descriptor):
                 event_data = event.data  # cache for perf
                 event_timestamps = event.timestamps
                 for field in discard_fields:
                     del event_data[field]
                     del event_timestamps[field]
-                for field in extra_fields:
+                if not no_fields_filter:
                     # Look in the descriptor, then start, then stop.
-                    if field in config_data:
+                    config_data = set(filter(comp_re.fullmatch, config_data)) - selected_fields
```

Comment:
Can't this be done outside of the loop over events? (above line 185)?

It looks like we should be able do to a 

```python
all_the_extra_data = {...}
all_the_extra_ts = {...}
for ev in mds.get_events_generator(...):
    ev['data'].update(all_the_extra_data)
    ev['timestamps'].update(all_the_extra_ts)
    yield ev
```